File: JugadorGanador.java
import java.util.Scanner;

public class JugadorGanador {
    public static void main(String[] args) {
        //String data = "5\n" +
        //        "140 82\n" +
        //        "89 134\n" +
        //        "90 110\n" +
        //        "112 106\n" +
        //        "88 90"; // Replace with your actual input data

        //se modifico para que la data venga de el archivo de texto y no de un texto hardcodeado
        String data ="";

        Scanner scanner = new Scanner(System.in);
        StringBuilder holder = new StringBuilder();

        while (scanner.hasNextLine()) {
            String line = scanner.nextLine();
            if (line.isEmpty()) {
                break; // Stop reading at an empty line
            }
            holder.append(line).append("\n");
        }
        data = holder.toString();

        scanner.close();

        //System.out.print(data);

        String[] lineas = data.trim().split("\n");

        int rondas = Integer.parseInt(lineas[0]);

        //declaro jugadores, cada jugador tendra:
        // su acumulado de puntuacion registro de la ventaja mas grande que han tenido al final de una ronda
        // y bandera que indica si son ganadores "so far"
        Jugador jug1 = new Jugador();
        Jugador jug2 = new Jugador();

        for (int rondaCont = 1; rondaCont <= rondas; rondaCont++) {
            String[] resRond = lineas[rondaCont].trim().split(" ");

            jug1.addPuntuacion(Integer.parseInt(resRond[0]));
            jug2.addPuntuacion(Integer.parseInt(resRond[1]));

            if (jug1.getPuntuacion() > jug2.getPuntuacion()) {
                verificarVentajaMayor(jug1, jug2);
            } else {
                verificarVentajaMayor(jug2, jug1);
            }
        }

        if (jug1.isGanando()) {
            System.out.println("1 " + jug1.getVentajaMayor());
        } else {
            System.out.println("2 " + jug2.getVentajaMayor());
        }
    }

    private static void verificarVentajaMayor(Jugador ganador, Jugador perdedor) {
        //Funcion que actualiza la ventaja major de el ganador, si es necesario
        int ventajaActual = ganador.getPuntuacion() - perdedor.getPuntuacion();
        if ( ganador.getVentajaMayor() < ventajaActual) {
            ganador.setVentajaMayor(ventajaActual);
            if (ventajaActual > perdedor.getVentajaMayor()) {
                ganador.setGanando(true);
                perdedor.setGanando(false);
            }
        }
    }
}

class Jugador {
    //Defino el objeto jugador que usare para registro de puntuacion y ganador
    private int puntuacion;
    private int ventajaMayor;
    private boolean ganando;

    public Jugador() {
        this.puntuacion = 0;
        this.ventajaMayor = 0;
        this.ganando = false;
    }


    //creacion de getter y setters para accesar a los parametros de Jugador
    public int getPuntuacion() {
        return puntuacion;
    }

    public void addPuntuacion(int puntuacionParam) {
        puntuacion += puntuacionParam;
    }

    public int getVentajaMayor() {
        return ventajaMayor;
    }

    public void setVentajaMayor(int ventajaMayorParam) {
        this.ventajaMayor = ventajaMayorParam;
    }

    public boolean isGanando() {
        return ganando;
    }

    public void setGanando(boolean ganandoParam) {
        this.ganando = ganandoParam;
    }
}

File: jugador_ganador.js
"use strict";
//Se agregaron validaciones,
Object.defineProperty(exports, "__esModule", { value: true });
exports.obtenerGanador = void 0;
var obtenerGanador = function (inputFilename, outputFilename) {
  if (outputFilename === void 0) {
    outputFilename = "jugador_ganador_type_results.txt";
  }
  var fs = require("fs");
  var lineas = new Array();
  var resultado = "";
  var errores = [];
  //regex para validar que las lineas solo contengan numeros y espacios vacios
  var regex = /^[0-9\s]+$/;
  try {
    lineas = fs.readFileSync(inputFilename, "utf-8").trim().split("\n");
    var rondas = parseInt(lineas[0].trim());
    //validacion de que primera contenga solo numeros
    if (!regex.test(lineas[0].trim())) {
      errores.push("la primera linea tiene characteres no permitidos");
    }
    //
    if (lineas.length - 1 === rondas) {
      //declaracion de objetos jugador, tiene la puntacion total, registro de la ventaja mas grande de cada jugador y bandera de si es el ganador "so far"
      var Jug1 = {
        puntuacion: 0,
        ventajaMayor: 0,
        ganando: false,
      };
      var Jug2 = {
        puntuacion: 0,
        ventajaMayor: 0,
        ganando: false,
      };
      var verificarVentajaMayor = function (ganador, perdedor) {
        if (ganador.ventajaMayor < ganador.puntuacion - perdedor.puntuacion) {
          //si la ventaja al terminar esta ronda es major a la guardada se actualiza registro de la mayor ventaja
          ganador.ventajaMayor = ganador.puntuacion - perdedor.puntuacion;
          //y se verifica quien es el ganador(el que lleva la mayor ventaja "so far")
          if (ganador.ventajaMayor > perdedor.ventajaMayor) {
            ganador.ganando = true;
            perdedor.ganando = false;
          }
        }
      };
      for (var rondaCont = 1; rondaCont <= rondas; rondaCont++) {
        //verificacion de errores, si hay errores salir del for
        if (errores.length === 0) {
          //iteramos por cada ronda
          //verificacion de que la linea para la ronda actual tenga caracteres permitidos
          if (!regex.test(lineas[rondaCont].trim())) {
            errores.push(
              "la linea para los resultados de la ronda " +
                rondaCont +
                " tiene characteres no permitidos"
            );
          }
          var resRond = lineas[rondaCont].trim().split(" ");
          //se agrega el resultado de la ronda al total de cada jugador
          Jug1.puntuacion = Jug1.puntuacion + parseInt(resRond[0]);
          Jug2.puntuacion = Jug2.puntuacion + parseInt(resRond[1]);
          //se comparan totales para ver quien va ganando despues de esta ultima ronda
          //el texto dice que se puede asumir que un jugador siempre gana pero de igual manera agregare validacion por si empatan jugadores
          if (Jug1.puntuacion === Jug2.puntuacion) {
            errores.push(
              "Error: Jugadores empataron al final de la ronda, no se permiten empates"
            );
          } else {
            if (Jug1.puntuacion > Jug2.puntuacion) {
              //una vez obtenido el jugador se revisa si la ventaja obtenida es la mayor
              verificarVentajaMayor(Jug1, Jug2);
            } else {
              //una vez obtenido el jugador se revisa si la ventaja obtenida es la mayor
              verificarVentajaMayor(Jug2, Jug1);
            }
          }
        } else {
          //si ya hubo errores se sale del for
          break;
        }
      }
      if (Jug1.ganando) {
        resultado = "1 " + Jug1.ventajaMayor;
        //console.log("1 " + Jug1.ventajaMayor);
      } else {
        resultado = "2 " + Jug2.ventajaMayor;
        //console.log("2 " + Jug2.ventajaMayor);
      }
    } else {
      //console.log("Numero de lineas incorrecto");
      errores.push("Numero de lineas incorrecto");
    }
  } catch (error) {
    errores.push(error.message);
    console.log("Error:", error.message);
  }
  //si hubo errores se imprimen los errores, si no, se imprime el resultado
  if (errores.length === 0) {
    fs.writeFileSync(outputFilename, resultado);
  } else {
    fs.writeFileSync(outputFilename, errores.join("\n"));
  }
};
exports.obtenerGanador = obtenerGanador;
var texto = "5\n  140 82\n  89 134\n  90 110\n  112 106\n  88 90";
//obtenerGanador();
if (require.main === module) {
  var inputFilename = process.argv[2];
  var outputFilename = process.argv[3];
  if (inputFilename) {
    if (outputFilename) {
      obtenerGanador(inputFilename, outputFilename);
    } else {
      obtenerGanador(inputFilename);
    }
  } else {
    //Mensaje Error Correcto
    console.log(
      "Uso: node jugador_ganador_type.js <input_filename> <output_filename> - (opcional)"
    );
  }
}
